libbpf-tools/task_detector.c

```diff
@@ -0,0 +1,376 @@
+			else {
+				time_to_str(ti.ts - p_start,
+						d_str, sizeof(d_str));
+				pr_ti(&ti, "DEQUEUE AFTER PREEMPTED", d_str);
+			}
+			break;
+		case TYPE_SYSCALL_ENTER:
+			siv = ti.ts;
+			sik.cpu = ti.cpu;
+			sik.pid = ti.pid;
+			sik.syscall = ti.syscall;
+			bpf_map_update_elem(env.si_map_fd, &sik, &siv, BPF_ANY);
+			syscall_name(ti.syscall, comm_buf, sizeof(comm_buf));
+			snprintf(func, sizeof(func), "SC [%d:%s] ENTER",
+					ti.syscall, comm_buf);
+			pr_ti(&ti, func, NULL);
+			break;
+		case TYPE_SYSCALL_EXIT:
+			sik.cpu = ti.cpu;
+			sik.pid = ti.pid;
+			sik.syscall = ti.syscall;
+			if (bpf_map_lookup_elem(env.si_map_fd, &sik, &siv))
+				break;
+			time_to_str(ti.ts - siv, d_str, sizeof(d_str));
+			bpf_map_delete_elem(env.si_map_fd, &sik);
+			syscall_name(ti.syscall, comm_buf, sizeof(comm_buf));
+			snprintf(func, sizeof(func), "SC [%d:%s] TAKE %s TO EXIT",
+					ti.syscall, comm_buf, d_str);
+			pr_ti(&ti, func, NULL);
+			break;
+		default:
+			break;
+		}
+
+		memcpy(&ti_last, &ti, sizeof(ti));
+	}
+
+	return end;
+}
+
+static int setup_user_info(int pid)
+{
+	int key = 0;
+	DIR *dir;
+	char buf[256];
+	struct user_info ui;
+	snprintf(buf, sizeof(buf), "/proc/%d", pid);
+	dir = opendir(buf);
+	if (!dir) {
+		printf("Open %s failed: %s\n", buf, strerror(errno));
+		return -1;
+	}
+	memset(&ui, 0, sizeof(ui));
+	ui.pid = pid;
+	ui.trace_syscall = env.trace_syscall;
+	bpf_map_update_elem(env.ui_map_fd, &key, &ui, BPF_ANY);
+
+	closedir(dir);
+
+	return 0;
+}
+
+static void int_exit(int sig)
+{
+	env.exiting = true;
+}
+
+int main(int argc, char **argv)
+{
+	struct task_detector_bpf *obj;
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	init_syscall_names();
+	
+	/* Check cpu number */
+	env.nr_cpus = sysconf(_SC_NPROCESSORS_CONF);
+	if (env.nr_cpus > NR_CPU_MAX) {
+		printf("Support Maximum %d cpus\n", NR_CPU_MAX);
+		goto freename;
+	}
+	
+	/* Increase rlimit */
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		goto freename;
+	}
+	
+	/* Open bpf object */
+	obj = task_detector_bpf__open();
+	if(!obj){
+		fprintf(stderr, "failed to open and/or load BPF object\n");
+		goto freename;
+	}
+
+	/* Load bpf program */
+	err = task_detector_bpf__load(obj);
+	if(err){
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+	
+	/* Attach bpf program */
+	err = task_detector_bpf__attach(obj);
+	if (err) {
+		fprintf(stderr, "failed to attach BPF programs\n");
+		goto cleanup;
+	}
+
+	/* Setup global map fd */
+	env.ui_map_fd = bpf_map__fd(obj->maps.user_info_maps);
+	env.ti_map_fd = bpf_map__fd(obj->maps.trace_info_maps);
+	env.si_map_fd = bpf_map__fd(obj->maps.syscall_info_maps);
+	env.start_map_fd = bpf_map__fd(obj->maps.start_maps);
+	env.end_map_fd = bpf_map__fd(obj->maps.end_maps);

```

Comment:
as I mentioned above, env is not intended to keep all programs state, for that just static global variables are a nicer alternative for such small tools. env is for what's parsed from command line arguments.